### PR TITLE
Refine new() and setup() methods

### PR DESCRIPTION
**Pull request by mulkieran**
This PR includes a bunch of changes.

* When making a new device still known to the kernel, it verifies that the given information matches the information known to the kernel.
* It removes the redundant size parameter from ```ThinPoolDev::new()``` and ```setup()```
* It makes running ```thin_check``` in ```ThinPoolDev::setup()``` conditional on the device being unknown to the kernel.
* It changes the return type of ```DM::parse_table_status()``` since the method prefers to panic in all cases where it might otherwise return an error.
* It trims trailing white space off the last entry in each line of the table returned by ```DM::parse_table_status()```.
* It add several tests to thinpooldev.rs.
* It does other small adjustements and fixes.

It addresses #120 but can not be certainly said to fix it, due to the uncertainty of what it means to be "in use".

### CODE REVIEW

**Comment by mulkieran** [comment]
@agrover Please take a look. The testing framework is incomplete and I don't intend to include it in the final PR. But it is handy right now.

**Comment by agrover** [comment]
Seems good so far.

Why do you think device list is more certain, in 34dbe4d?

Reloading the table if the device exists I hadn't thought of... I'll have to think more about this!

**Comment by mulkieran** [comment]
Other than using ```device_list()```, the only options are to try something, and then to assume that an error means the device does not exist. But an error could be caused by lots of reasons. Using ```device_list()``` seems less fuzzy. If there is an error, then we don't know anything. But if there is no error then the name is either in the list or it isn't. All very cut-and-dried.

**Comment by mulkieran** [comment]
About reloading the table...reloading or not reloading are equally fraught with danger, AFAICT. I can not see a way to interrogate DM to discover what the table contains. If there were then, in these ```new()``` or ```setup()``` methods as appropriate, I would instead do a comparison and fail if DM and Stratis view did not match up.

It may be that this is a question to raise with DM.

**Comment by agrover** [comment]
you can get current table by calling dm.table_status with the DM_STATUS_TABLE flag.

**Comment by mulkieran** [comment]
rebased.

**Comment by mulkieran** [comment]
rebased.

**Comment by mulkieran** [comment]
clippy has compilation errors, hopefully they will be fixed soon.

**Comment by mulkieran** [comment]
Making data block size the minimum value gets an I/O error instead of an invalid argument error. Progress.

**Comment by mulkieran** [comment]
@agrover @trgill review?

These are a bunch of disparate changes that are probably mostly correct. They do not completely solve any big problem, but they might as well go up for review now.

File: src/thindev.rs [hunk 1]
```diff
@@ -215,25 +216,24 @@ impl ThinDev {
...
     /// with its thin id from the thinpool.
     pub fn destroy(self, dm: &DM, thin_pool: &ThinPoolDev) -> DmResult<()> {
         let thin_id = self.thin_id;
         self.teardown(dm)?;
-        thin_pool.message(dm, &format!("delete {}", thin_id))?;
-
-        Ok(())
+        thin_pool
+            .message(dm, &format!("delete {}", thin_id))
+            .map(|_| ())
```
Comment (trgill) [inline src/thindev.rs]: the .map(|_| ()) is necessary here?

Comment (mulkieran) [inline src/thindev.rs]: I don't think it is. Removing.

**Comment by mulkieran** [comment]
Mad some changes to "Use a final map..." commit.

File: src/thindev.rs (hunk 1 above)
Comment (trgill) [inline src/thindev.rs]: FYI - there are a few cases.  I only marked this one.

Comment (mulkieran) [inline src/thindev.rs]: I looked over everything, the only changes are in "Use a final map..." commit. If there is anything there that you think is wrong, let me know.

**Comment by agrover** [comment]
Please drop the `map(|_| ())`s. They strike me as less clear than what's there now.

Remove `device_exists` check from `new()`. We want `setup` to work if the device already exists but I don't think this applies to `new`.

**Comment by mulkieran** [comment]
"Use expression in place of identifier" substituted for "Use final map...".

**Comment by mulkieran** [comment]
```ThinPoolDev::setup()``` invokes ```ThinPoolDev::new()```, so ```new()``` must work in the case that the device is already known to the kernel in order for ```setup()``` to work.

**Comment by mulkieran** [comment]
@agrover review?

**Comment by agrover** [comment]
> ThinPoolDev::setup() invokes ThinPoolDev::new(), so new() must work in the case that the device is already known to the kernel in order for setup() to work.

Here are the current requirements:

new:
1. doesn't call thin_check
1. errors out if name already exists

setup:
1. calls thin_check
1. doesn't error out if name already exists

Before, they only differed by requirement 1, so implementing one with the other was the simplest thing to do. Now that we have added requirement 2, this may no longer be the case, in which case the implementation must change, not the requirements.

**Comment by mulkieran** [comment]
How have these requirements been arrived at?

**Comment by agrover** [comment]
They were implicit in my comments on this PR. I wanted to explicitly state them since I'm not sure if there's a disagreement over the requirements or the implementation, and we can discuss them separately.

**Comment by mulkieran** [comment]
We would also like ability to teardown meta and data devices if construction of ThinPoolDev fails either in ```setup()``` or ```new()```, possibly. We should consider that complication, while we're worrying about all the rest.

I've pushed off the easily acked commits in this PR to #138.

**Comment by mulkieran** [comment]
rebased.

**Comment by mulkieran** [comment]
rebased.

**Comment by mulkieran** [comment]
rebased.

**Comment by mulkieran** [comment]
rebased.

**Comment by mulkieran** [comment]
```ThinPoolDev::new``` is called exactly when a pool is constructed for the first time. It is not clear whether it should call ```thin_check``` or not. It is reasonable, but maybe optimistic, to expect that invoking ```thin_check``` on a new thinpool would be benign, and might be useful.

**Comment by trgill** [comment]
@mulkieran if we use a recent version of "thin_check" it fails with:

syscall 'open' failed: Device or resource busy
Note: you cannot run this tool with these options on live metadata.
test real_test_pool_setup ... FAILED

It is why the CI tests are failing.

**Comment by mulkieran** [comment]
rebased.

**Comment by mulkieran** [comment]
@agrover In ```ThinPoolDev::new()``` why should an error be returned if the device already exists? Why can't we just continue if we happen to have the correct meta and data devices? Also, this bears on #153 .

**Comment by mulkieran** [comment]
rebased.

**Comment by agrover** [comment]
What's a use case for `new` not returning an error if the thinpool already exists? I can't think of one.

 If the user was setting up an existing thinpool they should've called `setup`. If the user actually is trying to create a new thinpool then returning an existing one will just cause weird failures down the line, because the user needs to track things like used thin_ids, and the assumption when `new` succeeds is that the thinpool is empty/unused, which will likely not be true if an existing thinpool is returned.

**Comment by mulkieran** [comment]
rebased.

**Comment by mulkieran** [comment]
rebased.

**Comment by mulkieran** [comment]
Obsoleted by a bunch of subsequent PRs.

**Pull request closed without merging**